### PR TITLE
Mafia finally grows docs

### PR DESCRIPTION
**Pull request by jacobstanley**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,340 @@
...
+using a nix-like hashing system so that for a given set of transitive
+dependencies, a package is only ever built once. Source dependencies are
+also cached, using their dependencies and source code as the hash. This
+is critical for Haskell development at scale. At Ambiata we have around
+150 Haskell packages which are being developed at a rapid pace, so it's
+crucial that we don't need to think about curating blessed sets of
+packages that we can cache as a single snapshot.
+
+
+Commands
```
Comment (charleso) [inline README.md:64]: Worth adding a section on using the `./mafai` file as well?

Comment (jacobstanley) [inline README.md:64]: Yeah I think so

File: README.md [hunk 2]
```diff
@@ -1,13 +1,340 @@
...
+having to resort to a monorepo.
+
+All dependencies are cached globally in `$HOME/.ambiata/mafia/packages`
+using a nix-like hashing system so that for a given set of transitive
+dependencies, a package is only ever built once. Source dependencies are
+also cached, using their dependencies and source code as the hash. This
+is critical for Haskell development at scale. At Ambiata we have around
+150 Haskell packages which are being developed at a rapid pace, so it's
+crucial that we don't need to think about curating blessed sets of
+packages that we can cache as a single snapshot.
```
Comment (charleso) [inline README.md:37]: Tempting to a fully fleshed out "motivation" section or separate doc, but then again maybe it's better not to...

File: README.md [hunk 3]
```diff
@@ -1,13 +1,340 @@
...
+
+
+### mafia bench
+
+Benchmark this project, by default this runs all the benchmarks.
+
+
+### mafia repl
+
+Start GHCi via `cabal repl`, by default on the library of the project.
```
Comment (charleso) [inline README.md:136]: Can you pass `test` to this?

File: README.md [hunk 4]
```diff
@@ -1,13 +1,340 @@
...
+
+
+### mafia quick
+
+Start GHCi directly on one or more files.
```
Comment (charleso) [inline README.md:141]: Worth adding a sentence on why this is better/different from `repl`?

**Comment by charleso** [comment]
:100:

File: README.md (hunk 3 above)
Comment (jacobstanley) [inline README.md:136]: I think so?

**Comment by thumphries** [comment]
Would be nice to mention the script, makes it pretty easy to get started.

Folks also might expect `mafia install` to install the package they're looking at

:shaved_ice: awesome

**Comment by jacobstanley** [comment]
> Folks also might expect `mafia install` to install the package in pwd

Yeah I was thinking a bit about that, it would also be awesome to be able to say:

```
$ $(mafia install ambiata/mismi/mismi-s3)/s3
```

and have it install executables from github repos

**Comment by jacobstanley** [comment]
@charleso @thumphries added a section on using the script

File: README.md [hunk 5]
```diff
@@ -1,13 +1,364 @@
...
+[script](https://raw.githubusercontent.com/ambiata/mafia/topic/docs/script/mafia)
+in the same directory as your `.cabal` project file and then run
+`./mafia upgrade`. This will ensure that you have the latest script and
+also bake the git hash of the Mafia version in to the script.
+
+You can now run `./mafia build` for the first time. This will build the
+Mafia executable with the git hash specified in the script, and then
+build the current project. Having the hash baked in to the script means
+that anyone trying to build the project will be using exactly the same
+version of Mafia.
```
Comment (charleso) [inline README.md:57]: :+1:

File: README.md (hunk 4 above)
Comment (jacobstanley) [inline README.md:141]: Hmm, not really sure what to say, ideally i'd like to get rid of repl and have quick be the only option if we can make it work in all the cases

File: README.md [hunk 6]
```diff
@@ -1,13 +1,364 @@
...
+to think about it.
+
+### mafia build
+
+Builds a project, including all executables, test suites and benchmarks.
+
+Mafia uses whichever GHC it finds on the `PATH` to build a project. It
+keeps Cabal sandboxes in version specific directories, so that it is
+easy to [switch](https://gist.github.com/jystic/efded8238bc4c64affb7)
+between versions of GHC while developing a project.
```
Comment (tmcgilchrist) [inline README.md:79]: This doesn't work so well with the default setup for emacs haskell-mode, it gets confused about the different versions for say 7.8 and 7.10.

**Comment by tmcgilchrist** [comment]
Worth mentioning the project.submodules file for pointing to non-git submodule dependencies.
@jystic

File: README.md (hunk 6 above)
Comment (jacobstanley) [inline README.md:79]: Yeah I have had similar problems with `ghc-mod` and `vim`, you have to always return to the ghc that was on the `PATH` when you started `vim` and then do a `mafia build` if you want `ghc-mod` to give useful information.

**Comment by jacobstanley** [comment]
@tmcgilchrist oh yeah, good point

**Comment by tmcgilchrist** [comment]
For those monorepo savages.

**Comment by jacobstanley** [comment]
@tmcgilchrist you know what, i think i might just fix it so you don't even need the .submodules file, probably easier than explaining its legacy.

**Comment by tmcgilchrist** [comment]
:rocket:

File: README.md [hunk 7]
```diff
@@ -1,13 +1,364 @@
 mafia
 =====
 
-> Provides protection against cabal swindling, robbing, injuring or sabotaging people with chopsticks.
+> Provides protection against cabal swindling, robbing, injuring or
+> sabotaging people with chopsticks.
 
 ![mafia](img/mafia.jpg)
```
Comment (damncabbage) [inline README.md:7]: Maybe ditch one or the other image. It's a bit scroll-heavy.

Comment (jacobstanley) [inline README.md:7]: Move it to the footer?

**Pull request merged**